src/txkube/testing/integration.py

```diff
@@ -102,14 +106,352 @@ def _named(kind, name, namespace=None):
+                return d
+            d.addBoth(cleanup)
+            return d
+        return wrapper
+    return decorator
+
+
+
+class _NamespaceTestsMixin(object):
+    @async
+    def test_namespace(self):
+        """
+        ``Namespace`` objects can be created and retrieved using the ``create``
+        and ``list`` methods of ``IKubernetesClient``.
+        """
+        obj = creatable_namespaces().example()
+        d = self.client.create(obj)
+        def created_namespace(created):
+            self.assertThat(created, matches_namespace(obj))
+            return self.client.list(v1.Namespace)
+        d.addCallback(created_namespace)
+        def check_namespaces(namespaces):
+            self.assertThat(namespaces, IsInstance(v1.NamespaceList))
+            # There are some built-in namespaces that we'll ignore.  If we
+            # find the one we created, that's sufficient.
+            self.assertThat(
+                namespaces.items,
+                AnyMatch(MatchesAll(matches_namespace(obj), has_uid(), is_active())),
+            )
+        d.addCallback(check_namespaces)
+        return d
+
+
+    @async
+    def test_namespace_retrieval(self):
+        """
+        A specific ``Namespace`` object can be retrieved by name using
+        ``IKubernetesClient.get``.
+        """
+        return self._global_object_retrieval_by_name_test(
+            creatable_namespaces(),
+            v1.Namespace,
+            matches_namespace,
+        )
+
+
+    @async
+    def test_namespace_deletion(self):
+        """
+        ``IKubernetesClient.delete`` can be used to delete ``Namespace``
+        objects.
+        """
+        obj = creatable_namespaces().example()
+        d = self.client.create(obj)
+        def created_namespace(created):
+            return self.client.delete(created)
+        d.addCallback(created_namespace)
+        def deleted_namespace(ignored):
+            return self.client.list(v1.Namespace)
+        d.addCallback(deleted_namespace)
+        def check_namespaces(collection):
+            active = list(
+                ns.metadata.name
+                for ns
+                in collection.items
+                if ns.status.phase == u"Active"
+            )
+            self.assertThat(
+                active,
+                Not(Contains(obj.metadata.name)),
+            )
+        d.addCallback(check_namespaces)
+        return d
+
+
+    @async
+    def test_duplicate_namespace_rejected(self):
+        """
+        ``IKubernetesClient.create`` returns a ``Deferred`` that fails with
+        ``KubernetesError`` if it is called with a ``Namespace`` object
+        with the same name as a *Namespace* which already exists.
+        """
+        return self._create_duplicate_rejected_test(
+            None, creatable_namespaces(), u"namespaces", None,
+        )
+
+
+
+class _ConfigMapTestsMixin(TestCase):
+    @async
+    @needs(namespace=creatable_namespaces().example())
```

Comment:
Oohh.  `data.draw`.  That seems to be the kind of thing I was looking for when I was first writing `needs`.  Yes, I think I will do something with that.  I'm filing a ticket for it now.